File: solutions/java/code-carl/ch09_dp/T6_LC0343_IntegerBreak.java
/**
 * Practice template for:
 * - LeetCode 343. Integer Break
 *   https://leetcode.cn/problems/integer-break/
 */
/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/20 23:47
 * @comment
 */
public class T6_LC0343_IntegerBreak {
    public int integerBreak(int n) {
        int[] dp = new int[n + 1];
        dp[0] = 0;
        dp[1] = 1;
        for (int i = 2; i <= n; i++) {
            for (int j = 1; j <= i / 2; j++) {
                dp[i] = Math.max(dp[i], Math.max(j * (i - j), j * dp[i - j]));
            }
        }
        return dp[n];
    }
}

File: solutions/python/code-carl/ch09_dp/t6_lc0343_integer_break.py
"""LeetCode 343. Integer Break
https://leetcode.cn/problems/integer-break/
"""
from typing import List


class Solution:
    def integerBreak(self, n: int) -> int:
        dp: List[int] = [0] * (n + 1)
        dp[0], dp[1], dp[2] = 0, 0, 1
        for i in range(3, len(dp)):
            for j in range(1, i // 2 + 1):
                dp[i] = max(dp[i], j * (i - j), j * dp[i - j])

        return dp[-1]


if __name__ == "__main__":
    solution = Solution()
    assert solution.integerBreak(2) == 1
    assert solution.integerBreak(10) == 36
    print("All tests passed.")
